rate-limiter/src/token_bucket.rs

```diff
@@ -86,55 +69,50 @@ where
     {
         TokenBucket {
             rate_per_second,
-            available: rate_per_second,
             requested: 0,
             last_update: now,
             time_provider,
         }
     }
 
-    fn calculate_delay(&self) -> Duration {
+    fn calculate_delay(&self) -> Option<Instant> {
         if self.rate_per_second == 0 {
-            return Duration::MAX;
+            return None;
         }
-        let delay_micros = (self.requested - self.available)
+        let delay_micros = (self.requested - self.rate_per_second)
             .saturating_mul(1_000_000)
             .saturating_div(self.rate_per_second);
-        Duration::from_micros(delay_micros.try_into().unwrap_or(u64::MAX))
+        Some(self.last_update + Duration::from_micros(delay_micros.try_into().unwrap_or(u64::MAX)))
     }
 
     fn update_units(&mut self, now: Instant) -> usize {
         let time_since_last_update = now.duration_since(self.last_update);
-        let mut new_units = time_since_last_update
+        self.last_update = now;
+        let new_units = time_since_last_update
             .as_micros()
             .saturating_mul(self.rate_per_second as u128)
             .saturating_div(1_000_000)
             .try_into()
             .unwrap_or(usize::MAX);
-        let used_new_units = min(new_units, self.requested);
-        new_units -= used_new_units;
-        self.requested -= used_new_units;
+        self.requested = self.requested.saturating_sub(new_units);
 
-        self.available = self.available.saturating_add(new_units);
-        self.last_update = now;
+        self.available()
+    }
 
-        let used = min(self.available, self.requested);
-        self.available -= used;
-        self.requested -= used;
-        self.available = min(self.available, self.token_limit());
-        self.available
+    fn available(&self) -> usize {
+        self.rate_per_second.saturating_sub(self.requested)
     }
 
     /// Calculates [Duration](time::Duration) by which we should delay next call to some governed resource in order to satisfy
     /// configured rate limit.
-    pub fn rate_limit(&mut self, mut requested: usize) -> Option<Duration> {
+    pub fn rate_limit(&mut self, requested: usize) -> Option<Option<Instant>> {
         trace!(
             target: LOG_TARGET,
             "TokenBucket called for {} of requested bytes. Internal state: {:?}.",
             requested,
             self
         );
-        if self.requested > 0 || self.available < requested {
+        if self.available() < requested {
```

Comment:
Can't we `update_units` first? This would simplify a bit the whole function.

Reply:
Changed. Current version is more clear.